src/section/section_map.rs

```diff
@@ -0,0 +1,461 @@
+                        && !pfx.is_extension_of(other_pfx)
+                };
+
+                if self.other.iter().any(is_newer) {
+                    return Some(*pfx);
+                }
+
+                None
+            })
+            .collect();
+
+        for pfx in to_remove {
+            let _ = self.other.remove(&pfx);
+        }
+    }
+
+    /// Returns the known section keys and any recent keys we still hold.
+    pub fn keys(&self) -> impl Iterator<Item = (&Prefix<XorName>, &SectionKeyInfo)> {
+        self.keys
+            .iter()
+            .chain(self.recent_keys.iter().map(|(p, k)| (p, k)))
+    }
+
+    pub fn latest_compatible_key(&self, name: &XorName) -> Option<&SectionKeyInfo> {
+        self.keys()
+            .filter(|(prefix, _)| prefix.matches(name))
+            .map(|(_, info)| info)
+            .max_by_key(|info| info.version())
+    }
+
+    /// Updates the entry in `keys` for `prefix` to the latest known key; if a split
+    /// occurred in the meantime, the keys for sections covering the rest of the address space are
+    /// initialised to the old key that was stored for their common ancestor
+    pub fn update_keys(&mut self, key_info: &SectionKeyInfo) {
+        trace!("attempts to update keys for {:?}", key_info);
+
+        if let Some((&old_pfx, old_version)) = self
+            .keys
+            .iter()
+            .find(|(pfx, _)| pfx.is_compatible(key_info.prefix()))
+            .map(|(pfx, info)| (pfx, info.version()))
+        {
+            if old_version >= key_info.version() || old_pfx.is_extension_of(key_info.prefix()) {
+                // Do not overwrite newer version or prefix extensions
+                return;
+            }
+
+            let old_key_info = self
+                .keys
+                .remove(&old_pfx)
+                .expect("Bug in BTreeMap for update_keys");
+
+            self.recent_keys.push_front((old_pfx, old_key_info.clone()));
+            if self.recent_keys.len() > MAX_RECENT_KEYS {
+                let _ = self.recent_keys.pop_back();
+            }
+
+            trace!("    from {:?} to {:?}", old_key_info, key_info);
+
+            let old_pfx_sibling = old_pfx.sibling();
+            let mut current_pfx = key_info.prefix().sibling();
+            while !self.keys.contains_key(&current_pfx) && current_pfx != old_pfx_sibling {
+                let _ = self.keys.insert(current_pfx, old_key_info.clone());
+                current_pfx = current_pfx.popped().sibling();
+            }
+        }
+        let _ = self.keys.insert(*key_info.prefix(), key_info.clone());
+    }
+
+    pub fn get_knowledge(&self, prefix: &Prefix<XorName>) -> Option<u64> {
+        self.knowledge.get(prefix).copied()
+    }
+
+    /// Returns the index of the public key in our_history that will be trusted by the target
+    /// location
+    pub fn proving_index(&self, target: &DstLocation) -> u64 {
```

Comment:
provable_index ?

Reply:
No, that wouldn't be accurate. But I'll rename it to something like `trusted_key_version` which is more explicit.